### PR TITLE
Message queue implementation

### PR DESCRIPTION
**Pull request by KinhTruongUnikie**
The program can be run only with the combination of --help(optional), --queue \<queue name\> and --file \<file name\>(in any order), any other combinations will result exit error. 
The main implementation sits in ipc_methods.c.

P.S short option e.g -h -q also work

### CODE REVIEW

File: C_hands_on_programming/ipc_methods.c [hunk 1]
```diff
@@ -323,17 +323,173 @@ void messageReceive(const ipc_info *info) {
...
+	buffer = (char*)malloc(size);
+
+	memset(&attrs, 0, sizeof attrs);
+	attrs.mq_maxmsg = 1;
+	attrs.mq_msgsize = size;
+	// open or create a mqueue if it does not exist
+	msg_queue = mq_open(info->argument_string, O_WRONLY | O_CREAT, 0666, &attrs);
+	if (msg_queue == -1) {
+		perror("msg_open");
+		exit(EXIT_FAILURE);
```
Comment (RomainChaumontet) [inline C_hands_on_programming/ipc_methods.c:432]: You shall free buffer before exiting (this comment is also available for the other EXIT_FAILURE)

**Review by RomainChaumontet — APPROVED** [review]
Minor comments.

I am not a fan about according the size of the queue  to the size of the file.

The help command doesn't say specifically to give some argument after the --queue option.

File: C_hands_on_programming/ipc_methods.c (hunk 1 above)
Comment (KinhTruongUnikie) [reply C_hands_on_programming/ipc_methods.c:432]: You are right, there are couple of places I overlooked the buffer not being freed, will fix that shortly

**Review by baovu-unikie — APPROVED** [review]

File: C_hands_on_programming/ipc_methods.c [hunk 2]
```diff
@@ -323,17 +323,173 @@ void messageReceive(const ipc_info *info) {
...
+	 size = lseek(fd, 0, SEEK_END);
+	 //set the file pointer back to beginning of file
+	 lseek(fd, current, SEEK_SET);
+
+	 // read file
+	 buffer = (char*)malloc(size);
+	 if (read(fd, buffer, size) == -1) {
+	     perror("read\n");
+	     exit(EXIT_FAILURE);
+	     free(buffer);
```
Comment (ivan-estonia) [inline C_hands_on_programming/ipc_methods.c:361]:
```
             perror("read\n");
	     free(buffer);
	     close(fd);
	     exit(EXIT_FAILURE);
	     ```

File: C_hands_on_programming/ipc_methods.c [hunk 3]
```diff
@@ -323,17 +323,173 @@ void messageReceive(const ipc_info *info) {
...
+	 }
+	 close(fd);
+	 //write to pipe
+	 if (write(fdp, buffer, size) == -1) {
+	     perror("write\n");
+	     exit(EXIT_FAILURE);
+	     free(buffer);
```
Comment (ivan-estonia) [inline C_hands_on_programming/ipc_methods.c:368]: `exit()` should be the last

File: C_hands_on_programming/ipc_methods.c [hunk 4]
```diff
@@ -323,17 +323,173 @@ void messageReceive(const ipc_info *info) {
...
+	 // open file for writing
+	 fd = open(info->filename, O_WRONLY | O_CREAT, 0644);//mode(file permission should be specified if O_CREATE is used)
+	 if (fd == -1) {
+		 perror("open");
+		 exit(EXIT_FAILURE);
+	 }
+	 // read from pipe and write to file
+	 buffer = (char*)malloc(PIPE_BUF);
+	 while ((bytes = read(fdp, buffer, PIPE_BUF)) != 0) {
+		 write(fd, buffer, bytes);
```
Comment (ivan-estonia) [inline C_hands_on_programming/ipc_methods.c]: /* TODO: error handler: bytes == -1 */

File: C_hands_on_programming/ipc_methods.c [hunk 5]
```diff
@@ -323,17 +323,173 @@ void messageReceive(const ipc_info *info) {
...
 void queueReceive(const ipc_info *info){
+	printf("Starting queueReceive..\n");
+	struct mq_attr attrs;
+	int n;
+	char *buffer = NULL;
+	mqd_t msg_queue;
+	FILE *writeFile;
+	// open or create a mqueue if it does not exist
+	printf("Looking for the message queue:\n");
+	while ((msg_queue = mq_open(info->argument_string, O_RDONLY)) == -1);
```
Comment (ivan-estonia) [inline C_hands_on_programming/ipc_methods.c]: You should add a delay to this cycle or it will block the processor core.

**Review by ivan-estonia — CHANGES_REQUESTED** [review]

File: C_hands_on_programming/ipc_methods.c (hunk 2 above)
Comment (KinhTruongUnikie) [reply C_hands_on_programming/ipc_methods.c:361]: Fixed now

File: C_hands_on_programming/ipc_methods.c (hunk 3 above)
Comment (KinhTruongUnikie) [reply C_hands_on_programming/ipc_methods.c:368]: Fixed

File: C_hands_on_programming/ipc_methods.c (hunk 4 above)
Comment (KinhTruongUnikie) [reply C_hands_on_programming/ipc_methods.c]: Fixed

File: C_hands_on_programming/ipc_methods.c (hunk 2 above)
Comment (ivan-estonia) [reply C_hands_on_programming/ipc_methods.c:361]: `close(fd);`

**Review by ivan-estonia — APPROVED** [review]

**Pull request merged**